### PR TITLE
MDGPS

### PR DESCRIPTION
**Pull request by wmontgomery4**
Sorry for the delay, here's a preliminary pull request for the mdgps code. It isn't in a state to be merged just yet, but I wanted to let you take a look at the main outline. The standing issues right now are:
1) Some of the changes are pretty hacky (particularly to get on-policy sampling working), and should be made more sane before merging. This also includes adding a separate traj_opt file for MDGPS, and the way that I use a 'pol_info' object as a 'traj_distr' object sometimes.
2) I didn't change any of the tf policy_opt code, but did make changes to the caffe policy_opt code, so things probably aren't compatible between the two right now
3) On-policy sampling isn't working that well with just 4 conditions on the peg insertion task (we used 9 in the paper), so I need to play around with that and see if it's just a hyperparameter issue.
4) There are also some minor tweaks to the gui and some fixed TODOs from non-mdgps stuff, which I should probably just open a new branch for

I'm working on the issues above, but feel free to take a look over things in the meantime.

### CODE REVIEW

**Comment by cbfinn** [comment]
Thanks Harley! I'll take a look at it now, and go ahead and repost/comment when you make changes that are ready to be looked at.

File: mjc_models/obstacle_course.xml [hunk 1]
```diff
@@ -0,0 +1,41 @@
+<mujoco model="linearparticle">
```
Comment (cbfinn) [inline mjc_models/obstacle_course.xml]: Since there are now a fair number of xml files, it would be nice to have comments in this file and/or a screen shot of the environment for each xml file.

File: python/gps/algorithm/algorithm_mdgps.py [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+
+    def _update_step_size(self):
+        """ Evaluate costs on samples, and adjust the step size. """
+        # Evaluate cost function for all conditions and samples.
+        for m in range(self.M):
+            self._update_policy_fit(m, init=True)
+            self._eval_cost(m)
+            lmax = self.cur[m].cs.max(axis=0)
+            lmin = self.cur[m].cs.min(axis=0)
+            self.cur[m].qmax = np.cumsum( (lmax - lmin)[::-1] )[::-1]
```
Comment (cbfinn) [inline python/gps/algorithm/algorithm_mdgps.py]: what is this code doing? [maybe add a explanatory comment?]

File: python/gps/algorithm/algorithm_mdgps.py [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+
+            # Adjust step size relative to the previous iteration.
+            if self.iteration_count > 0:
+                self._stepadjust(m)
+
+    def _update_policy(self, itr, inner_itr):
+        """ Compute the new policy. """
```
Comment (cbfinn) [inline python/gps/algorithm/algorithm_mdgps.py:102]: It would be nice if we could reuse more code between this file and algorithm_badmm. (Not a super high priority though)

File: python/gps/algorithm/traj_opt/traj_opt_lqr_python.py [hunk 4]
```diff
@@ -244,7 +253,7 @@ def backward(self, prev_traj_distr, traj_info, eta, algorithm, m):
                     fail = True
                     break
 
-                # Store conditional covariance, inverse, and Cholesky.
+                # Store conditional covariance, inverse, and Cholesky
```
Comment (cbfinn) [inline python/gps/algorithm/traj_opt/traj_opt_lqr_python.py]: nitpic: periods are nice (and are consistent with pylint standards).

File: python/gps/algorithm/traj_opt/traj_opt_utils.py [hunk 5]
```diff
@@ -85,7 +85,7 @@ def traj_distr_kl(new_mu, new_sigma, new_traj_distr, prev_traj_distr):
         )
 
     # Add up divergences across time to get total divergence.
-    return np.sum(kl_div)
+    return kl_div
```
Comment (cbfinn) [inline python/gps/algorithm/traj_opt/traj_opt_utils.py]: It seems like this change would break something in traj_opt_lqr.py... Was this a bug in the code?

Comment (wmontgomery4) [inline python/gps/algorithm/traj_opt/traj_opt_utils.py]: No, this was a mistake I made when setting up the pull request, along with several of the other comments. I had some code supporting weighted KL constraints but took it out since it wasn't helping and cluttered things. I'm making a fix for that now.

File: python/gps/algorithm/algorithm_mdgps.py (hunk 2 above)
Comment (wmontgomery4) [inline python/gps/algorithm/algorithm_mdgps.py]: This was for an idea Sergey had where we compute the cumulative Qmax (like in the bound from the paper), in order to try and inform the step size updates. We never got it working, and I forgot to remove it.

**Comment by wmontgomery4** [comment]
Okay, this is in a state to be looked at again. Couple notes:
- Works with tf as well, but it's slower and doesn't seem to work quite as well (although this could easily be a fluke, or because the hyperparams were tweaked with caffe). Tensorflow also takes up a lot of memory, so I can't run all 4 variations simultaneously, so I'm sticking with caffe for now, but it works with tensorflow and so that's not an issue.
- I had to remove the default TrajOptLqrPython from the algorithm config.py to avoid circular imports (I need to import both AlgorithmBADMM and AlgorithmMDGPS in TrajOpt for some type checks since the traj opt varies slightly between them)
- Still haven't refactored things with BADMM, but I agree that would be a good next step.

File: python/gps/algorithm/algorithm_mdgps.py [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+            self.cur[m].pol_info.policy_prior = \
+                    policy_prior['type'](policy_prior)
+
+        self.policy_opt = self._hyperparams['policy_opt']['type'](
+            self._hyperparams['policy_opt'], self.dO, self.dU
+        )
+
+    def iteration(self, sample_lists):
+        """
+        Run iteration of BADMM-based guided policy search.
```
Comment (cbfinn) [inline python/gps/algorithm/algorithm_mdgps.py]: change BADMM to mirror descent

File: python/gps/algorithm/algorithm_mdgps.py [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+        self._update_policy_samples()  # Choose samples to use with the policy.
+
+        # On the first iteration we need to make sure that the policy somewhat
+        # matches the init controller. Otherwise the LQR backpass starts with
+        # a bad linearization, and things don't work out well.
+        if self.iteration_count == 0:
+            self.new_traj_distr = [
+                self.cur[cond].traj_distr for cond in range(self.M)
+            ]
+            self._update_policy(self.iteration_count, 1) # HACK: need to set inner_itr=1
```
Comment (cbfinn) [inline python/gps/algorithm/algorithm_mdgps.py]: Why is this a hack? I guess it would be really easy to change the policy_opt_caffe code to instead take in a normalize=True/False flag, or check if the policy scale and bias already exists. Could you make this change? [It will also involve a small change to algorithm_badmm]

File: python/gps/algorithm/algorithm_mdgps.py [hunk 8]
```diff
@@ -0,0 +1,334 @@
...
+
+        # Main updates
+        # TODO: store init_kl after updating the trajectory
+        self._update_step_size()  # KL Divergence step size (also fits policy).
+        self._update_trajectories()
+        self._update_policy(self.iteration_count, 1) # HACK: need to set inner_itr=1
```
Comment (cbfinn) [inline python/gps/algorithm/algorithm_mdgps.py]: why do you set inner_itr to 1 here? (why recompute the mean and variance?)

File: python/gps/algorithm/algorithm_mdgps.py [hunk 9]
```diff
@@ -0,0 +1,334 @@
...
+        # Store most recent kl divs for GUI output
+        for m in range(self.M):
+            kl_m = self._policy_kl(m)[0]
+            self.cur[m].pol_info.prev_kl = kl_m
+
+        self._advance_iteration_variables()
+
+    def _update_policy_samples(self):
+        """ Update the list of samples to use with the policy. """
+        #TODO: Handle synthetic samples.
```
Comment (cbfinn) [inline python/gps/algorithm/algorithm_mdgps.py]: Do you plan to add synthetic samples? If not, let's remove this TODO.

File: python/gps/algorithm/config.py [hunk 10]
```diff
@@ -18,13 +15,13 @@
     'init_traj_distr': None,  # A list of initial LinearGaussianPolicy
                               # objects for each condition.
     # Trajectory optimization.
-    'traj_opt': TrajOptLQRPython({}),
```
Comment (cbfinn) [inline python/gps/algorithm/config.py:21]: I think this is okay. In practice, we'll always specify this in the hyperparams anyway.

File: python/gps/algorithm/config.py [hunk 11]
```diff
@@ -44,3 +41,12 @@
     'exp_step_upper': 0.5,
     'exp_step_lower': 1.0,
 }
+
+# AlgorithmMD
+ALG_MDGPS = {
+    'inner_iterations': 4,
```
Comment (cbfinn) [inline python/gps/algorithm/config.py]: Looks like this is unused. Remove it?

File: python/gps/algorithm/config.py [hunk 12]
```diff
@@ -44,3 +41,12 @@
...
+    'init_pol_wt': 0.01,
```
Comment (cbfinn) [inline python/gps/algorithm/config.py:48]: Is this also unused?

File: python/gps/gps_main.py [hunk 13]
```diff
@@ -19,7 +19,7 @@
 from gps.utility.data_logger import DataLogger
 from gps.sample.sample_list import SampleList
 
-logging.basicConfig(format='%(levelname)s:%(message)s', level=logging.INFO)
```
Comment (cbfinn) [inline python/gps/gps_main.py:22]: might be nice to keep the default here as INFO.

File: python/gps/algorithm/traj_opt/traj_opt_lqr_python.py [hunk 14]
```diff
@@ -30,7 +32,19 @@ def update(self, m, algorithm):
         step_mult = algorithm.cur[m].step_mult
         prev_eta = algorithm.cur[m].eta
         traj_info = algorithm.cur[m].traj_info
-        prev_traj_distr = algorithm.cur[m].traj_distr
+
+        if type(algorithm) == AlgorithmMDGPS:
+            # For MDGPS, constrain to previous NN linearization
+            # NOTE: we only copy in necessary params, ignoring inv_pol_covar
+            pol_info = algorithm.cur[m].pol_info
```
Comment (cbfinn) [inline python/gps/algorithm/traj_opt/traj_opt_lqr_python.py]: might be nice to put this in a helper function. It's rather verbose.

File: python/gps/algorithm/traj_opt/traj_opt_lqr_python.py [hunk 15]
```diff
@@ -221,13 +239,20 @@ def backward(self, prev_traj_distr, traj_info, eta, algorithm, m):
...
                 # Add in the value function from the next time step.
                 if t < T - 1:
-                    #TODO: Should multiply by
-                    #      (pol_wt[t+1] + eta)/(pol_wt[t] + eta) here.
-                    Qtt = Qtt + \
-                            Fm[t, :, :].T.dot(Vxx[t+1, :, :]).dot(Fm[t, :, :])
-                    Qt = Qt + \
-                            Fm[t, :, :].T.dot(Vx[t+1, :] +
-                                              Vxx[t+1, :, :].dot(fv[t, :]))
+                    if type(algorithm) == AlgorithmBADMM:
```
Comment (cbfinn) [inline python/gps/algorithm/traj_opt/traj_opt_lqr_python.py:236]: Can we just write this as something like:

```
if type(algorithm) == AlgortihmBADMM:
    multiplier = (pol_wt[t+1] + eta)/(pol_wt[t] + eta)
else:
    multiplier = 1
Qtt = Qtt + multiplier * ...
```

I think this would be easier to read, and slightly less code. If you can think of a better variable name for multiplier, feel free to change.

**Comment by cbfinn** [comment]
Generally looks good! Left some fairly minor comments. 

I'm think it would be okay to merge before refactoring with BADMM, and leave the refactor for the next PR, since this one is already rather large.

File: python/gps/algorithm/algorithm_mdgps.py (hunk 7 above)
Comment (wmontgomery4) [inline python/gps/algorithm/algorithm_mdgps.py]: It's a bit of a hack because it's not really inner_itr 1, you just have to say that so that the mean and variance will be set correctly. Agreed that adding a normalize flag is the better option, and I'll make that change. In the future we might be better off just using batch norm, especially since the conditions will be changing randomly, but let's put that off for now.

File: python/gps/algorithm/algorithm_mdgps.py (hunk 8 above)
Comment (wmontgomery4) [inline python/gps/algorithm/algorithm_mdgps.py]: Think this was just a c&p error, no need to recompute the mean/variance as you say.

File: python/gps/algorithm/algorithm_mdgps.py (hunk 9 above)
Comment (wmontgomery4) [inline python/gps/algorithm/algorithm_mdgps.py]: No, this was copied and pasted from the BADMM code, should I remove the TODO there as well?

File: python/gps/algorithm/config.py (hunk 12 above)
Comment (wmontgomery4) [inline python/gps/algorithm/config.py:48]: Yes, these are both unused, thanks for the catch!

File: python/gps/gps_main.py (hunk 13 above)
Comment (wmontgomery4) [inline python/gps/gps_main.py:22]: Ok no problem. I've found DEBUG pretty useful, but I can just keep that as a local change.

File: python/gps/algorithm/traj_opt/traj_opt_lqr_python.py (hunk 14 above)
Comment (wmontgomery4) [inline python/gps/algorithm/traj_opt/traj_opt_lqr_python.py]: I was debating about that. My plan was to just add a `pol_info_to_traj_distr` function in `algorithm_utils.py`, but I decided against it because the proper thing would be to compute 'inv_pol_covar' and that's never needed. Definitely premature optimization though.. so I'll just add that helper method. In the refactor, I think we might want to add a "linearization" object which can come from a policy or LG controller. We could use that everywhere traj_distr or a neural net linearization is needed.

File: python/gps/algorithm/traj_opt/traj_opt_lqr_python.py (hunk 15 above)
Comment (wmontgomery4) [inline python/gps/algorithm/traj_opt/traj_opt_lqr_python.py:236]: Sure, sounds good.

File: python/gps/algorithm/config.py (hunk 12 above)
Comment (wmontgomery4) [inline python/gps/algorithm/config.py:48]: Actually, I still need 'init_pol_wt', because the pol_info constructor requires it (and the policy_opt code uses pol_wt in several places). It doesn't matter for mdgps since the pol_wt remains constant, but it would be a good thing to separate out when refactor.

File: python/gps/algorithm/algorithm_mdgps.py (hunk 7 above)
Comment (wmontgomery4) [inline python/gps/algorithm/algorithm_mdgps.py]: Ended up just checking whether or not the scale/bias exists already. Seemed like an easier option for now, which could be replaced with batch norm later.

**Comment by wmontgomery4** [comment]
Okay, I've addressed the comments and switched the hyperparams file over to using tensorflow. I need to tweak the hyperparams a bit, but everything should be essentially working.

One note, I changed the kl_div_i/f outputs to show the average kl_div rather than the sum, since this is what the `step` column corresponds to already. If you'd prefer the sum let me know and I'll change it back.

Also, I had a bizarre bug where tensorflow would seg fault unless I continued to import `policy_opt_caffe.py` (even though it was unused). I fixed it by changing the order of imports in `policy_opt_tf.py`, but I'm still not sure what the issue was (it might just have been a local issue).

File: python/gps/algorithm/algorithm_mdgps.py (hunk 7 above)
Comment (cbfinn) [inline python/gps/algorithm/algorithm_mdgps.py]: Sounds good. I'd rather use a variant of batch norm which does not rely on the statistics of the batch, since we need to evaluate single samples effectively. There are a couple of options in this direction.

File: python/gps/algorithm/algorithm_mdgps.py (hunk 9 above)
Comment (cbfinn) [inline python/gps/algorithm/algorithm_mdgps.py]: Let's keep it in BADMM for now. I'm not sure if we'll add it for BADMM, probably not, but nice to be aware of.

File: python/gps/algorithm/config.py (hunk 12 above)
Comment (cbfinn) [inline python/gps/algorithm/config.py:48]: Ok, sounds good. Maybe just add a note that it's algorithmically unused.

File: python/gps/algorithm/traj_opt/traj_opt_lqr_python.py (hunk 14 above)
Comment (cbfinn) [inline python/gps/algorithm/traj_opt/traj_opt_lqr_python.py]: Ok, sounds good!

**Comment by cbfinn** [comment]
Ok, mean kl is fine. That's a rather weird tf bug.

I'll add you as a collaborator so you can merge whenever you're ready.
